### PR TITLE
Always visit conversions in array initializers

### PR DESCRIPTION
**Pull request by 333fred**
We were skipping visiting conversions in array initializers when no best type was found; this could potentially suppress some false positives in error scenarios, but it also meant that we weren't visiting expressions that expect to only be visited through a conversion, such as lambdas with a function type. Closes https://github.com/dotnet/roslyn/issues/72056.

### CODE REVIEW

File: eng/build.sh [hunk 1]
```diff
@@ -26,6 +26,7 @@ usage()
   echo "Test actions:"
   echo "  --testCoreClr              Run unit tests on .NET Core (short: --test, -t)"
   echo "  --testMono                 Run unit tests on Mono"
+  echo "  --testCompilerOnly         Run only the compiler unit tests"
```
Comment (333fred) [inline eng/build.sh:29]: This is a duplicate of the same flag that we already have for `build.ps1`. Separated into a separate commit, and I will be doing a rebase merge to keep it separate.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -61862,6 +61862,35 @@ static void M()
...
+                }
+                """, options: WithNullableEnable(TestOptions.ReleaseExe));
+
+            comp.VerifyDiagnostics(
+                // (6,11): error CS0103: The name 'M2' does not exist in the current context
+                //     () => M2(e),
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "M2").WithArguments("M2").WithLocation(6, 11),
+                // (7,14): warning CS8604: Possible null reference argument for parameter 'param' in 'void Program.M1(IEnumerable<string> param)'.
+                //     () => M1(e)
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "e").WithArguments("param", "void Program.M1(IEnumerable<string> param)").WithLocation(7, 14)
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:61890]: Before the change, this warning was not reported, and we hit an assert in `NullableWalker.DebugVerifier` that the `M1(e)` `BoundCall` node was not visited during analysis.

**Review by RikkiGibson — COMMENTED** [review]
> this could potentially suppress some false negatives

Do we have any test which shows a difference in cascading diagnostics before and after this change?

**Review by RikkiGibson — APPROVED** [review]

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for the second review.

**Review by cston — APPROVED** [review]

**Pull request merged**